File: js/common/PeekIterator.js
const LinkedList = require('linkedlist');

class PeekIterator {

    CACHE_SIZE=10;
    it;
    // 需要putback的元素
    stackPutBacks;
    // 基于时间差UN个Ian的缓存
    queueCache;
    enToken;
    constructor(it,endToken=null) {
        this.it=it
        this.stackPutBacks = new LinkedList()
        this.queueCache = new LinkedList()

        this.enToken = endToken
    }


    peek(){
        if (this.stackPutBacks.length > 0){
            return this.stackPutBacks.head
        }

        // peek 相当于 先 next 出来在放回去
        const val = this.next()
        this.putBack()
        return val
    }

    putBack(){
        if (this.queueCache.length>0){
            this.stackPutBacks.push(this.queueCache.pop())
        }
    }

    hasNext(){
        return this.enToken || !!this.peek()
    }

    next(){
        let val ;
        if (this.stackPutBacks.length>0){
            val = this.stackPutBacks.pop()
        }else{
            val = this.it.next().value
            if (val===undefined){

                const tmp = this.enToken
                this.enToken=null
                return tmp
            }
        }

        // 处理缓存
        while(this.queueCache.length > this.CACHE_SIZE -1){
            this.queueCache.shift()
        }

        this.queueCache.push(val)

        return val

    }

}

module.exports=PeekIterator

File: java/src/main/java/common/PeekIterator.java
package common;

import java.util.Iterator;
import java.util.LinkedList;
import java.util.stream.Stream;

/**
 * 把字符串变为可迭代
 * @param <T>
 */
public class PeekIterator<T> implements Iterator<T> {

    private static final int CACHE_SIZE = 10;
    private final Iterator<T> it;
    private final LinkedList<T> queueCache = new LinkedList<>();
    private final LinkedList<T> statPutBacks = new LinkedList<>();
    private T _endToken;


    public PeekIterator(Stream<T> stream) {
        it = stream.iterator();
    }
    public PeekIterator(Stream<T> stream, T endToKen) {
        it = stream.iterator();
        _endToken = endToKen;

    }

    public T peek() {
        if (this.statPutBacks.size() > 0) {
            return this.statPutBacks.getFirst();
        }
        if (!it.hasNext()) {
            return _endToken;
        }
        T val = next();
        putBack();
        return val;
    }

    /**
     * A -> B -> C -> D
     * D -> C -> B -> A
     */
    public void putBack() {
        if (queueCache.size() > 0) {
            this.statPutBacks.push(queueCache.pollLast());
        }
    }

    @Override
    public boolean hasNext() {
        return  _endToken!=null ||
                this.statPutBacks.size()>0 ||
                it.hasNext();
    }

    @Override
    public T next() {

        T val;

        if (this.statPutBacks.size() > 0) {
            val = this.statPutBacks.pop();
        } else {
            if (!it.hasNext()){
                T tmp = _endToken;
                _endToken=null;
                return tmp;
            }
            val = it.next();
        }

        // 缓存最近的10个 , 这里缓存 9个
        while (queueCache.size() > CACHE_SIZE - 1) {
            // 过大则移除
            queueCache.poll();
        }
        // 在最后添加 1 个
        queueCache.add(val);

        return val;
    }
}
